### PR TITLE
[opt](scan) Use lazy-init for segment iterators and avoid caching all segments in the rowset reader

### PR DESCRIPTION
**Pull request by mrhhsg**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by mrhhsg** [comment]
run buildall

File: be/src/olap/rowset/beta_rowset_reader.cpp [hunk 1]
```diff
@@ -388,11 +394,7 @@ bool BetaRowsetReader::_should_push_down_value_predicates() const {
 }
 
 Status BetaRowsetReader::get_segment_num_rows(std::vector<uint32_t>* segment_num_rows) {
-    auto& seg_ptrs = _segment_cache_handle.get_segments();
-    segment_num_rows->resize(seg_ptrs.size());
-    for (size_t i = 0; i < seg_ptrs.size(); i++) {
-        (*segment_num_rows)[i] = seg_ptrs[i]->num_rows();
-    }
+    segment_num_rows->assign(_segments_rows.cbegin(), _segments_rows.cend());
```
Comment (yiguolei) [inline be/src/olap/rowset/beta_rowset_reader.cpp:397]: could you make sure that this method is called after get_segment_iterators???

File: be/src/olap/rowset/segment_v2/segment.cpp [hunk 2]
```diff
@@ -132,6 +131,14 @@ Status Segment::_open_inverted_index() {
 
 Status Segment::new_iterator(SchemaSPtr schema, const StorageReadOptions& read_options,
                              std::unique_ptr<RowwiseIterator>* iter) {
+    auto st = _create_column_readers_once.call([&] {
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/segment.cpp:134]: Should also check in other methods that will use column readers.

File: be/src/olap/rowset/beta_rowset_reader.cpp [hunk 3]
```diff
@@ -249,38 +250,60 @@ Status BetaRowsetReader::get_segment_iterators(RowsetReaderContext* read_context
...
+                return Status::Error<ROWSET_READER_INIT>(status.to_string());
+            }
         } else {
-            DCHECK_EQ(seg_end - seg_start, _segment_row_ranges.size());
-            auto local_options = _read_options;
-            local_options.row_ranges = _segment_row_ranges[i - seg_start];
-            status = seg_ptr->new_iterator(_input_schema, local_options, &iter);
+            if (_segment_row_ranges.empty()) {
+                _read_options.row_ranges.clear();
+                iter = std::make_unique<LazyInitSegmentIterator>(seg_ptr, _input_schema,
```
Comment (yiguolei) [inline be/src/olap/rowset/beta_rowset_reader.cpp:296]: Add some comment here to explain why we need lazy init segment iterator.

**Comment by mrhhsg** [comment]
run buildall

File: be/src/olap/rowset/beta_rowset_reader.cpp [hunk 4]
```diff
@@ -249,38 +250,66 @@ Status BetaRowsetReader::get_segment_iterators(RowsetReaderContext* read_context
...
+
     auto [seg_start, seg_end] = _segment_offsets;
     if (seg_start == seg_end) {
         seg_start = 0;
         seg_end = segments.size();
     }
 
+    const bool is_merge_iterator = _is_merge_iterator();
+    const bool use_lazy_init_iterators =
+            !is_merge_iterator && _read_context->reader_type == ReaderType::READER_QUERY;
```
Comment (yiguolei) [inline be/src/olap/rowset/beta_rowset_reader.cpp:272]: why need check reader type?

File: be/src/olap/rowset/beta_rowset_reader.cpp [hunk 5]
```diff
@@ -388,11 +417,7 @@ bool BetaRowsetReader::_should_push_down_value_predicates() const {
...
 
 Status BetaRowsetReader::get_segment_num_rows(std::vector<uint32_t>* segment_num_rows) {
-    auto& seg_ptrs = _segment_cache_handle.get_segments();
-    segment_num_rows->resize(seg_ptrs.size());
-    for (size_t i = 0; i < seg_ptrs.size(); i++) {
-        (*segment_num_rows)[i] = seg_ptrs[i]->num_rows();
-    }
+    segment_num_rows->assign(_segments_rows.cbegin(), _segments_rows.cend());
     return Status::OK();
 }
```
Comment (yiguolei) [inline be/src/olap/rowset/beta_rowset_reader.cpp:422]: some method like new_column_iterator also need load column readers

**Comment by mrhhsg** [comment]
run buildall

**Review by yiguolei — APPROVED** [review]

**Review by yiguolei — APPROVED** [review]

**Review by jacktengg — APPROVED** [review]

**Pull request merged**